### PR TITLE
Add RouteBuilder::nameExists()

### PR DESCRIPTION
**Pull request by mirko-pagliai**
See https://github.com/cakephp/cakephp/issues/9312

### CODE REVIEW

File: src/Routing/Router.php [hunk 1]
```diff
@@ -727,6 +727,16 @@ public static function reverse($params, $full = false)
...
     }
+    
+    /**
+     * Checks if there is already a route with a given name.
+     * @param string $name Name.
+     * @return bool
+     */
+    public static function nameExists($name)
+    {
+        return in_array($name, array_keys(static::$_collection->named()));
```
Comment (markstory) [inline src/Routing/Router.php]: Couldn't this be `array_key_exists($name, static::$_collection->named());`

Comment (mirko-pagliai) [inline src/Routing/Router.php]: @markstory you're right

**Comment by markstory** [comment]
Should this method be on `RoutesBuilder` instead of `Router`? I'm trying to avoid adding more static methods to router in favour of building routes inside scoped builders. Right now the usage would read a bit inconsistently.

```php
Router::scope('/', function ($routes) {
  if (!Router::nameExists('login')) {
    $routes->connect(...);
  }
});

Whereas on the builder, the scopes read a bit more consistently:

```php
Router::scope('/', function ($routes) {
  if (!$routes->nameExists('login')) {
    $routes->connect(...);
  }
});
```

**Comment by saeideng** [comment]
or maybe,change exist method 
//RouteBuilder
function named(name=null){
return $_named[named]
or return false if not exist that key
finaally, return all named route if $name arg =null
@markstory ?

File: tests/TestCase/Routing/RouterTest.php [hunk 2]
```diff
@@ -1199,10 +1195,6 @@ public function testNamedRouteExists()
             ['action' => 'view'],
             ['_name' => 'Articles::view']
         );
-        
-        $this->assertTrue(Router::nameExists('users-index'));
-        $this->assertTrue(Router::nameExists('test'));
-        $this->assertTrue(Router::nameExists('Articles::view'));
```
Comment (markstory) [inline tests/TestCase/Routing/RouterTest.php]: Do these tests do anything still?

Comment (mirko-pagliai) [inline tests/TestCase/Routing/RouterTest.php]: @markstory What?

Comment (markstory) [inline tests/TestCase/Routing/RouterTest.php]: This test only connects a few routes and makes no assertions, seems like it isn't worth keeping around.

Comment (mirko-pagliai) [inline tests/TestCase/Routing/RouterTest.php]: OMG, I forgot that I have added the entire method, lol :D

File: tests/TestCase/Routing/RouterTest.php [hunk 3]
```diff
@@ -1174,7 +1174,7 @@ public function testUrlGenerationNamedRoute()
         $url = Router::url(['_name' => 'Articles::view', '1', '#' => 'frag']);
         $this->assertEquals('/view/1#frag', $url);
     }
-
+
```
Comment (dereuromark) [inline tests/TestCase/Routing/RouterTest.php]: Please make sure your IDE is "trim right" configured.

File: tests/TestCase/Routing/RouteBuilderTest.php [hunk 4]
```diff
@@ -168,6 +168,22 @@ public function testConnectTrimTrailingSlash()
...
+     * @return void
+     */
+    public function testNameExists()
+    {
+        $routes = new RouteBuilder($this->collection, '/l', ['prefix' => 'api']);
+        
+        $this->assertFalse($routes->nameExists('myRouteName'));
+        
+        $routes->connect('myRouteUrl', ['action' => 'index'], ['_name' => 'myRouteName']);
+
```
Comment (dereuromark) [inline tests/TestCase/Routing/RouteBuilderTest.php]: Same here, whitespace noise gets introduced here :(

File: src/Routing/RouteBuilder.php [hunk 5]
```diff
@@ -195,6 +195,16 @@ public function params()
     {
         return $this->_params;
     }
+
```
Comment (dereuromark) [inline src/Routing/RouteBuilder.php]: and here

**Comment by jadb** [comment]
@mirko-pagliai maybe squash the commits?

File: src/Routing/RouteBuilder.php [hunk 6]
```diff
@@ -197,6 +197,16 @@ public function params()
     }
 
     /**
+     * Checks if there is already a route with a given name.
+     * @param string $name Name.
```
Comment (ADmad) [inline src/Routing/RouteBuilder.php]: There should be an empty line above

**Comment by mirko-pagliai** [comment]
@jadb Maybe I'm a noob :-D
How to squash commits?

**Comment by ADmad** [comment]
@mirko-pagliai http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by saeideng** [comment]
hi
@markstory what you think about named() method?
```
//RouteBuilder
function named($name=null){
if(!empty($name) && isset([$_named[$named]]) ){
     return $_named[named];
}
 else if(!empty($name) ){return false;}
else{return $_named}
```

**Comment by markstory** [comment]
@saeideng I'm not a fan of how many return value types that method has.

**Comment by lorenzo** [comment]
This looks good to me

**Pull request merged**